src/D2L.CodeStyle.Analyzers/Helpers/ConstantAnalyzer.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Helpers {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class ConstantAnalyzer : DiagnosticAnalyzer {
+		private const string AttributeName = "D2L.CodeStyle.Annotations.Contract.ConstantAttribute";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Diagnostics.NonConstantPassedToConstantParameter,
+				Diagnostics.InvalidConstantType
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		public static void CompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeParameters(
+					ctx,
+					(ParameterListSyntax)ctx.Node
+				),
+				SyntaxKind.ParameterList
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					( (InvocationExpressionSyntax)ctx.Node ).ArgumentList
+				),
+				SyntaxKind.InvocationExpression
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					( (ObjectCreationExpressionSyntax)ctx.Node ).ArgumentList
+				),
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private static void AnalyzeParameters(
+			SyntaxNodeAnalysisContext context,
+			ParameterListSyntax parameterList
+		) {
+			SeparatedSyntaxList<ParameterSyntax> parameters = parameterList.Parameters;
+			foreach( ParameterSyntax parameterSyntax in parameters ) {
+				// We need the symbol
+				IParameterSymbol parameter = context.SemanticModel.GetDeclaredSymbol( parameterSyntax );
+
+				// Parameter is somehow null, so do nothing
+				if( parameter == null ) {
+					continue;
+				}
+
+				// Parameter is not [Constant], so do nothing
+				if( !parameter.HasAttribute( AttributeName ) ) {
+					continue;
+				}
+
+				// Get the base type of the parameter
+				ITypeSymbol type = parameter.Type;
+
+				// Check that the parameter type is not a valid type,
+				// and that it is not a type parameter
+				if( type.SpecialType != SpecialType.None
+				 || type.Kind != SymbolKind.NamedType ) {

```

Comment:
nit: formatting
```suggestion
				if( type.SpecialType != SpecialType.None
					 || type.Kind != SymbolKind.NamedType
	 	 	 	 ) {
```

But also, I don't think I understand the `type.Kind != SymbolKind.NamedType`. Could you explain?

Reply:
I split those into two separate statements so that I could add more explanation.
Basically the same reason as the argument block you asked about, a "NamedType" means it's not generic.